### PR TITLE
[PUBDEV-4865] Fix several issues affecting H2O client reconnection

### PR DESCRIPTION
**Pull request by jakubhava**
@michalkurka This PR is fixing the client reconnections issues consistetly acros flatfile and multicast mode. Please see the comments as it was easier to comment relevant parts of the code directly

### CODE REVIEW

File: h2o-core/src/main/java/water/Paxos.java [hunk 1]
```diff
@@ -70,7 +70,13 @@ static synchronized int doHeartbeat( H2ONode h2o ) {
         // A new client was reported to this node so we propagate this information to all nodes in the cluster, to this
         // as well
         UDPClientEvent.ClientEvent.Type.CONNECT.broadcast(h2o);
-      } else if (H2O.ARGS.client && !H2O.isNodeInFlatfile(h2o)) {
+      } else if (!H2O.ARGS.client && h2o._heartbeat._client) {
```
Comment (jakubhava) [inline h2o-core/src/main/java/water/Paxos.java]: When we put the ip:port of the client node to the flatfile passed to the nodes, this node is not added to the client map and therefore it's not handled properly when it comes to disconnects. We need to ensure that client specified this way needs to be on the client map as well

File: h2o-core/src/main/java/water/H2O.java [hunk 2]
```diff
@@ -2177,6 +2180,7 @@ public static H2ONode reportClient(H2ONode client) {
   }
 
   public static H2ONode removeClient(H2ONode client){
+    client.stopSendThread();
```
Comment (jakubhava) [inline h2o-core/src/main/java/water/H2O.java]: Stop the send thread of the client which was disconnected

File: h2o-core/src/main/java/water/H2ONode.java [hunk 3]
```diff
@@ -145,7 +145,16 @@ public boolean isHealthy(long now) {
   // *interned*: there is only one per InetAddress.
   static private H2ONode intern( H2Okey key ) {
     H2ONode h2o = INTERN.get(key);
-    if( h2o != null ) return h2o;
+    if( h2o != null){
```
Comment (jakubhava) [inline h2o-core/src/main/java/water/H2ONode.java]: When the client connects for the first time, this `intern` method is called and `h2o.new UDP_TCP_SendThread()` is started for the new client. When the client disconnects and connects again we need to start the sendThread manually again as it was stopped for the disconnecting client

File: h2o-core/src/main/java/water/H2O.java [hunk 4]
```diff
@@ -2042,6 +2042,9 @@ public static void main( String[] args ) {
     // Log registered parsers
     Log.info("Registered parsers: " + Arrays.toString(ParserService.INSTANCE.getAllProviderNames(true)));
 
+    // Start thread checking client disconnections
+    new ClientDisconnectCheckThread().start();
```
Comment (jakubhava) [inline h2o-core/src/main/java/water/H2O.java]: Start the thread which is checking whether client should be disconnected or not

File: h2o-core/src/main/java/water/FailedNodeWatchdogExtension.java [hunk 5]
```diff
@@ -221,7 +194,18 @@ public void run() {
             while (true) {
                 for(H2ONode client: H2O.getClients()){
                     if(isTimeoutExceeded(client, watchdogClientRetryTimeout)){
-                        handleClientDisconnect(client, watchdogClientRetryTimeout);
+                        // if timeout exceeded, check if the client is disconnected
```
Comment (jakubhava) [inline h2o-core/src/main/java/water/FailedNodeWatchdogExtension.java]: In watchdog client mode, we additionally need to check that all nodes agree on the fact that the client was disconected, however, the check mechanism whether the client disappeared from the single node needs to be performed also in the non-watchdog mode to handle the client disconnections properly consistently.

File: h2o-core/src/main/java/water/ClientDisconnectCheckThread.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
+package water;
```
Comment (mmalohlava) [inline h2o-core/src/main/java/water/ClientDisconnectCheckThread.java]: Is this another listener like `ListenerService`?

File: h2o-core/src/main/java/water/ClientDisconnectCheckThread.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+  @Override
+  public void run() {
+    while (true) {
+      for(H2ONode client: H2O.getClients()){
+        if(isTimeoutExceeded(client, HeartBeatThread.CLIENT_TIMEOUT * 2)){
+          handleClientDisconnect(client);
+        }
+      }
+      try {
+        Thread.sleep(HeartBeatThread.CLIENT_TIMEOUT * 2);
```
Comment (mmalohlava) [inline h2o-core/src/main/java/water/ClientDisconnectCheckThread.java]: Why `*2` ?

File: h2o-core/src/main/java/water/ClientDisconnectCheckThread.java (hunk 6 above)
Comment (jakubhava) [reply h2o-core/src/main/java/water/ClientDisconnectCheckThread.java]: I wouldn't say this is a listener. This is a basically a simple thread checking whether the client has been disconnected or not.

But we might actually store the history of joined/disconnected clients using the listenerservice to keep track of it.

File: h2o-core/src/main/java/water/ClientDisconnectCheckThread.java (hunk 7 above)
Comment (jakubhava) [reply h2o-core/src/main/java/water/ClientDisconnectCheckThread.java]: It doesn't mean anything special - just wait 2 times more than is the regular client timeout. Probably good to have this configurable, with some decent default value

Comment (jakubhava) [reply h2o-core/src/main/java/water/ClientDisconnectCheckThread.java]: Made it configurable

**Comment by jakubhava** [comment]
Thanks @mmalohlava for the review. What do you think pls@michalkurka?

File: h2o-core/src/main/java/water/H2ONode.java [hunk 8]
```diff
@@ -145,7 +145,16 @@ public boolean isHealthy(long now) {
...
+      if(h2o._heartbeat._client && h2o._removed_from_cloud){
+        // the client has reconnected, we need to restore the state
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/H2ONode.java]: it would be good to have this in a method (eg. startSendThread) close to the method `stopSendThread`

Comment (jakubhava) [reply h2o-core/src/main/java/water/H2ONode.java]: good point, will do!

File: h2o-core/src/main/java/water/ClientDisconnectCheckThread.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+    return (System.currentTimeMillis() - client._last_heard_from) >= timeout;
+  }
+
+  /**
+   * This method checks whether the client is disconnected from this node due to some problem such as client or network
+   * is unreachable.
+   */
+  private void handleClientDisconnect(H2ONode node) {
+    if(node != H2O.SELF) {
+      Log.warn("Client " + node + " disconnected!");
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/ClientDisconnectCheckThread.java]: This code is similar but not the same as UDPClientEvent.DISCONNECT - what is the reason for the difference?

File: h2o-core/src/main/java/water/H2O.java [hunk 10]
```diff
@@ -2048,6 +2059,9 @@ public static void main( String[] args ) {
...
```
Comment (michalkurka) [inline h2o-core/src/main/java/water/H2O.java]: Is this thread started even for the client itself?

File: h2o-core/src/main/java/water/ClientDisconnectCheckThread.java (hunk 9 above)
Comment (jakubhava) [reply h2o-core/src/main/java/water/ClientDisconnectCheckThread.java]: `UDPClientEvent.DISCONNECT` is the event which is triggered on purpose, usually, the clients decide to disconnect. In that case we also need to make sure that if this client is the watchdog client, we need to stop the cluster.

These cases are not handled here on purpose as this class is to handle client disconnections due to timeouts etc. It is up to the WatchDogClientExtension to handle correspondingly on the event that the watchdog client disappeared.

File: h2o-core/src/main/java/water/H2O.java (hunk 10 above)
Comment (jakubhava) [reply h2o-core/src/main/java/water/H2O.java]: yes, it is started for each node. It seems easier to handle each node the same than handling client and non-clients nodes differently. What do you think?

File: h2o-core/src/main/java/water/H2ONode.java (hunk 8 above)
Comment (jakubhava) [reply h2o-core/src/main/java/water/H2ONode.java]: done

**Comment by jakubhava** [comment]
Thank you all @michalkurka @mmalohlava !!

**Pull request merged**